src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,401 @@
+                        {
+                            builder.Add(parsedAttribute);
+                        }
+                    }
+                }
+            }
+
+            if (builder.Count > 0)
+            {
+                platformSpecificOperations.Add(operation, builder.ToImmutable());
+            }
+        }
+
+        private static bool IsWithinConditionalOperation(IOperation pOperation) =>
+            pOperation.ConstantValue.HasValue &&
+            pOperation.Parent is IBinaryOperation bo &&
+            (bo.OperatorKind == BinaryOperatorKind.Equals ||
+            bo.OperatorKind == BinaryOperatorKind.NotEquals ||
+            bo.OperatorKind == BinaryOperatorKind.GreaterThan ||
+            bo.OperatorKind == BinaryOperatorKind.LessThan ||
+            bo.OperatorKind == BinaryOperatorKind.GreaterThanOrEqual ||
+            bo.OperatorKind == BinaryOperatorKind.LessThanOrEqual);
+
+        private static bool IsOSPlatformsEqual(string left, string right)
+        {
+            if (left.Equals(right, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+            else
+            {
+                var normalizedLeft = NormalizeOSName(left);
+                var normalizedRight = NormalizeOSName(right);
+                return string.Equals(normalizedLeft, normalizedRight, StringComparison.OrdinalIgnoreCase);
+            }
+        }
+
+        private static string NormalizeOSName(string name) => name.Equals(OSX, StringComparison.OrdinalIgnoreCase)
+            ? macOS
+            : name;
+
+        private static DiagnosticDescriptor SelectRule(PlatformAttributeType attributeType)
+            => attributeType switch
+            {
+                PlatformAttributeType.MinimumOSPlatformAttribute => MinimumOsRule,
+                PlatformAttributeType.ObsoletedInOSPlatformAttribute => ObsoleteRule,
+                _ => RemovedRule,
+            };
+
+        private static ImmutableArray<AttributeData> FindAllPlatformAttributesApplied(ImmutableArray<AttributeData> immediateAttributes, INamedTypeSymbol parent)
+        {
+            using var builder = ArrayBuilder<AttributeData>.GetInstance();
+            foreach (AttributeData attribute in immediateAttributes)
+            {
+                if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name))
+                {
+                    builder.Add(attribute);
+                }
+            }
+
+            while (parent != null)
+            {
+                var current = parent.GetAttributes();
+                foreach (var attribute in current)
+                {
+                    if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name) &&
+                        TargetPlatformAttribute != attribute.AttributeClass.Name)
+                    {
+                        builder.Add(attribute);
+                    }
+                }
+                parent = parent.BaseType;
```

Comment:
The attributes defined at https://github.com/dotnet/designs/blob/master/accepted/2020/platform-checks/platform-checks.md#attributes indicate none of these attributes have `Inherited = true`, which means attributes applied to base types are not inherited by derived types. Did you actually mean to walk the `ContainingType` chain instead of base types? @terrajobst to confirm

```suggestion
                parent = parent.ContainingType;
```

Reply:
Actually it seems that you can also apply all of these attributes at assembly level with global assembly attributes. So, you either need to handle assembly symbol attributes specially after this loop OR might be easier to just walk the `ContainingSymbol` chain instead of `ContainingType` chain, which will include the `IAssemblySymbol` automatically.

Reply:
> Did you actually mean to walk the ContainingType chain instead of base types

That is right, mean to walk the ContainingType, thanks

> Actually it seems that you can also apply all of these attributes at assembly level with global assembly attributes. So, you either need to handle assembly symbol attributes specially after this loop OR might be easier to just walk the ContainingSymbol chain instead of ContainingType chain, which will include the IAssemblySymbol automatically.

Yes, attributes could be at the assembly level, i thought it would be covered with looping through all ContainingType chain. Because of I cannot have different assemblies in one source for testing i haven't yet run any test for assembly-level attributes (planned to do that when we can consume preview 8 for testing).

Reply:
> i thought it would be covered with looping through all ContainingType chain

ContainingType chain only walks outwards covering all types. You instead want to walk the ContainingSymbol chain.

> Because of I cannot have different assemblies in one source for testing

You do not need this for testing assembly attributes. You should be able to write tests with `[assembly: XXXPlatformAttribute(...)]` in your input source and they should work.

Reply:
> You do not need this for testing assembly attributes. You should be able to write tests with [assembly: XXXPlatformAttribute(...)] in your input source and they should work.

That could partially work, but wouldn't show full functionality as if the source has assembly-level attribute that would apply to all members inside and no warning for calling any member within it. I need to call a member from another assembly to test full functionality.